pkg/templates/usertiers/usertier_generator.go

```diff
@@ -0,0 +1,256 @@
+package usertiers
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	toolchainv1alpha1 "github.com/codeready-toolchain/api/api/v1alpha1"
+	"github.com/codeready-toolchain/host-operator/pkg/templates/assets"
+	commonclient "github.com/codeready-toolchain/toolchain-common/pkg/client"
+	commonTemplate "github.com/codeready-toolchain/toolchain-common/pkg/template"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/kubernetes/scheme"
+
+	templatev1 "github.com/openshift/api/template/v1"
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var log = logf.Log.WithName("usertiers")
+
+// CreateOrUpdateResources generates the UserTier resources,
+// then uses the manager's client to create or update the resources on the cluster.
+func CreateOrUpdateResources(s *runtime.Scheme, client client.Client, namespace string, assets assets.Assets) error {
+
+	// initialize tier generator, loads templates from assets
+	generator, err := newUserTierGenerator(s, client, namespace, assets)
+	if err != nil {
+		return errors.Wrap(err, "unable to create UserTier generator")
+	}
+
+	// create the UserTier resources
+	err = generator.createUserTiers()
+	if err != nil {
+		return errors.Wrap(err, "unable to create UserTiers")
+	}
+
+	return nil
+}
+
+type tierGenerator struct {
+	client          client.Client
+	namespace       string
+	scheme          *runtime.Scheme
+	templatesByTier map[string]*tierData
+}
+
+type tierData struct {
+	name         string
+	rawTemplates *templates
+	objects      []client.Object
+}
+
+// templates: any resources belonging to a given user tier ("advanced", "base", "team", etc.), currently only a single template
+type templates struct {
+	userTier *template // UserTier resource with tier-scoped configuration
+}
+
+// template: a template's content and its latest git revision
+type template struct {
+	revision string
+	content  []byte
+}
+
+// newUserTierGenerator loads templates from the provided assets and processes the UserTiers
+func newUserTierGenerator(s *runtime.Scheme, client client.Client, namespace string, assets assets.Assets) (*tierGenerator, error) {
+	// load templates from assets
+	templatesByTier, err := loadTemplatesByTiers(assets)
+	if err != nil {
+		return nil, err
+	}
+
+	c := &tierGenerator{
+		client:          client,
+		namespace:       namespace,
+		scheme:          s,
+		templatesByTier: templatesByTier,
+	}
+
+	// process UserTiers
+	if err := c.initUserTiers(); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// loadTemplatesByTiers loads the assets and dispatches them by tiers, assuming the given `assets` has the following structure:
+//
+// metadata.yaml
+// advanced/
+//   tier.yaml
+// basic/
+//   tier.yaml
+// team/
+//   tier.yaml
```

Comment:
maybe update the names to match what we have in the directory

Reply:
Done in 85d8b484e5d0b8a28a92e116d1508f16babcf3f6